include/realizations/catchment/Bmi_Formulation.hpp

```diff
@@ -0,0 +1,233 @@
+         *
+         * @param soil_m
+         * @return ``0``
+         */
+        double calc_et(double soil_m) override {
+            return 0;
+        }
+
+        const vector<std::string> &get_required_parameters() override {
+            return REQUIRED_PARAMETERS;
+        }
+
+    protected:
+
+        const bool &get_allow_model_exceed_end_time() const {
+            return allow_model_exceed_end_time;
+        }
+
+        const string &get_bmi_init_config() const {
+            return bmi_init_config;
+        }
+
+        const string &get_bmi_main_output_var() const {
+            return bmi_main_output_var;
+        }
+
+        /**
+         * Get the backing model object implementing the BMI.
+         *
+         * @return Shared pointer to the backing model object that implements the BMI.
+         */
+        std::shared_ptr<M> get_bmi_model() {
+            return bmi_model;
+        }
+
+        const string &get_forcing_file_path() const {
+            return forcing_file_path;
+        }
+
+        /**
+         * Get the name of the specific type of the backing model object.
+         *
+         * @return The name of the backing model object's type.
+         */
+        std::string get_model_type_name() {
+            return model_type_name;
+        };
+
+        /**
+         * Get the values making up the header line from get_output_header_line(), but organized as a vector of strings.
+         *
+         * @return The values making up the header line from get_output_header_line() organized as a vector.
+         */
+        const vector<std::string> &get_output_header_fields() const {
+            return output_header_fields;
+        }
+
+        /**
+         * Get the names of variables in formulation output.
+         *
+         * Get the names of the variables to include in the output from this formulation, which should be some ordered
+         * subset of the output variables from the model.
+         *
+         * @return
+         */
+        const vector<std::string> &get_output_variable_names() const {
+            return output_variable_names;
+        }
+
+        bool is_bmi_uses_forcing_file() const {
+            return bmi_uses_forcing_file;
+        }
+
+        /**
+         * Test whether the backing model object has been initialize using the BMI standard ``Initialize`` function.
+         *
+         * @return Whether backing model object has been initialize using the BMI standard ``Initialize`` function.
+         */
+        virtual bool is_model_initialized() {
+            return model_initialized;
+        };
+
+        void set_allow_model_exceed_end_time(bool allow_exceed_end) {
+            allow_model_exceed_end_time = allow_exceed_end;
+        }
+
+        void set_bmi_init_config(const string &init_config) {
+            bmi_init_config = init_config;
+        }
+
+        void set_bmi_main_output_var(const string &main_output_var) {
```

Comment:
Not entirely sure I'm following you.  

The idea behind this is that NGen expects/needs there to be a "main" BMI output variable.  This is where the value returned by the formulation's `get_response` function comes from.  The notion is completely specific to NGen.

Because we can't know variable names at compile-time for all models, and because of the way we access BMI variable values via the API, we need to be able to store and access this "main" BMI variable name we want `get_response` to use.  `bmi_main_output_var` might not be the best member variable name to use, so I am open to suggestions on that.

Reply:
What I was trying to say is `main_output_var` should be more explicit, i.e. `set_bmi_catchment_flow_var` or something that is more informative of the intent/specifics of ngen.

Reply:
So first, I've been thinking about this in context of both the config file interface (where `main_output_variable` is set) and the `Bmi_Formulation` class.

How carefully do we need to consider the potential implicit restrictions/expectations on the behavior of `get_response` and the model itself?  Perhaps something like `bmi_response_output_var`?